### PR TITLE
Improve angle transformation in `MottonenStatePreparation`

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
In `MottonenStatePreparation`, there is a step where the desired phases need to be transformed into rotation angles for the Gray code-based rotation sequence that implements the phases.
The sequence is shown in Fig. 2, the angle transformation in Eq. (3) of [Möttönen et al.](https://arxiv.org/pdf/quant-ph/0407010).

Currently, the transformation is done by explicitly constructing the matrix $M$ from Eq.(3) and multiplying it to the angles.
This is not only painstakingly slow but also requires quadratic memory size compared to the input and output. If we care about preparing larger state vectors, this limits the computation to about 15 qubits, which it does not have to do.

Instead, one can compute the angles with a fast Walsh-Hadamard transform (FWHT) and subsequent permutations. These permutations act as if we would apply a ladder of CNOT gates to the angles, when interpreting them as a quantum state.

**Description of the Change:**
Change `compute_theta`, which implements the transform, from the matrix-based method to the  FWHT-based one.

**Benefits:**
Roughly reduces memory requirement from `O(d^2)` to `O(d)`, where `d` is the input (and output) dimension.
Reduces compute time _heavily_:
image

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

### CODE REVIEW

**Comment by josh146** [comment]
😍

File: pennylane/templates/subroutines/select_pauli_rot.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def __init__(
         self.hyperparameters["target_wire"] = qml.wires.Wires(target_wire)
         self.hyperparameters["rot_axis"] = rot_axis
 
-        if qml.math.shape(angles)[0] != 2 ** len(control_wires):
+        if qml.math.shape(angles)[-1] != 2 ** len(control_wires):
```
Comment (dwierichs) [inline pennylane/templates/subroutines/select_pauli_rot.py:93]: This was preventing broadcasting of `SelectPauliRot`, although its methods all seem to be fine with broadcasting :D

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 2]
```diff
@@ -35,6 +35,40 @@ def test_standard_validity():
     qml.ops.functions.assert_valid(op)
 
 
+def compute_theta_reference(alpha):
```
Comment (dwierichs) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:38]: This is the old implementation (up to a speed up due to numpy broadcasting).

File: tests/templates/test_subroutines/test_select_pauli_rot.py [hunk 3]
```diff
@@ -130,15 +130,37 @@ def test_correctness(self, angles, rot_axis):
 
         assert np.isclose(1.0, output[0])
 
-    def test_decomposition(self):
+    @pytest.mark.parametrize("n", [1, 2, 3, 4])
+    @pytest.mark.parametrize("axis", "XYZ")
+    @pytest.mark.parametrize("batch_dim", [None, 1, 3])
+    def test_decomposition(self, n, axis, batch_dim):
```
Comment (dwierichs) [inline tests/templates/test_subroutines/test_select_pauli_rot.py:136]: Refined this test a little to include other rotation axis, qubit counts and broadcasting. Strictly speaking only the last addition tests code that has been changed in this PR.

File: pennylane/templates/state_preparations/mottonen.py [hunk 4]
```diff
@@ -14,83 +14,107 @@
...
-        (float): transformation matrix entry at given row and column
+        np.ndarray[int]: Array of ``2**rank`` integers that make up the Gray code.
     """
-    # (col >> 1) ^ col is the Gray code of col
-    b_and_g = row & ((col >> 1) ^ col)
-    sum_of_ones = 0
-    while b_and_g > 0:
-        if b_and_g & 0b1:
-            sum_of_ones += 1
+    g = np.array([0, 1])
```
Comment (albi3ro) [inline pennylane/templates/state_preparations/mottonen.py:35]: I was actually just looking at this code and wanting to  make a similar change too 😄 Happy this is being done. Why create strings just to convert them to ints 🤯

Comment (dwierichs) [reply pennylane/templates/state_preparations/mottonen.py:35]: Haha, funny! :D Yeah, it's really unnecessary!

File: pennylane/templates/state_preparations/mottonen.py [hunk 5]
```diff
@@ -14,83 +14,107 @@
...
 
-    return (-1) ** sum_of_ones
+_walsh_hadamard_matrix = np.array([[1, 1], [1, -1]]) / 2
+_cnot_matrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]]).reshape((2,) * 4)
 
 
-def compute_theta(alpha):
-    """Maps the angles alpha of the multi-controlled rotations decomposition of a uniformly controlled rotation
-     to the rotation angles used in the Gray code implementation.
+def compute_theta(D: TensorLike, n: Optional[int] = None):
```
Comment (albi3ro) [inline pennylane/templates/state_preparations/mottonen.py:45]: Ok if we rename `n` to `num_qubits`? Not quite sure if there's a better variable name for `D`, but at least we can fix `n`.

File: pennylane/templates/state_preparations/mottonen.py [hunk 6]
```diff
@@ -132,19 +156,17 @@ def _apply_uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
...
 
     code = gray_code(gray_code_rank)
-    num_selections = len(code)
-
-    control_indices = [
-        int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % num_selections], 2)))
-        for i in range(num_selections)
-    ]
-
+    control_indices = np.log2(code ^ np.roll(code, -1)).astype(int)
```
Comment (albi3ro) [inline pennylane/templates/state_preparations/mottonen.py:159]: ❤️

**Comment by albi3ro** [comment]
🚀

File: pennylane/templates/state_preparations/mottonen.py [hunk 7]
```diff
@@ -132,19 +156,17 @@ def _apply_uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
...
+
+    # For abstract or differentiated theta we will never skip a rotation. Likewise if there is at
+    # least one non-zero angle (per batch if batched) for all rotations.
+    nonzero = (theta != 0.0) if qml.math.ndim(theta) == 1 else qml.math.any(theta != 0.0, axis=0)
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/templates/state_preparations/mottonen.py:163]: It won't :facepalm:

[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/templates/state_preparations/mottonen.py:163]: Fixed.

**Review by albi3ro — APPROVED** [review]

File: pennylane/templates/state_preparations/mottonen.py [hunk 8]
```diff
@@ -14,83 +14,107 @@
...
-def compute_theta(alpha):
-    """Maps the angles alpha of the multi-controlled rotations decomposition of a uniformly controlled rotation
-     to the rotation angles used in the Gray code implementation.
+def compute_theta(alpha: TensorLike, num_qubits: Optional[int] = None):
+    r"""Maps the angles alpha of the multi-controlled rotations decomposition of a
+    uniformly controlled rotation to the rotation angles used in the Gray code implementation.
+    This function uses the fact that the transformation given by Eq. (3) in
+    `Möttönen et al. (2004) <https://arxiv.org/abs/quant-ph/0407010>`_ is equal to a Walsh-Hadamard
+    transform followed by some permutations, which can be expressed as a ladder of CNOT gates
+    applied to the angles, when interpreting them as a quantum state.
```
Comment (KetpuntoG) [inline pennylane/templates/state_preparations/mottonen.py:51]: What do you mean here in "when interpreting them as a quantum state."?

File: pennylane/templates/state_preparations/mottonen.py [hunk 9]
```diff
@@ -132,19 +156,19 @@ def _apply_uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
...
-        int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % num_selections], 2)))
-        for i in range(num_selections)
-    ]
-
+    control_indices = np.log2(code ^ np.roll(code, -1)).astype(int)
+
+    # For abstract or differentiated theta we will never skip a rotation. Likewise if there is at
+    # least one non-zero angle (per batch if batched) for all rotations.
+    skip_none = qml.math.is_abstract(theta) or qml.math.requires_grad(theta)
+    if not skip_none:
```
Comment (KetpuntoG) [inline pennylane/templates/state_preparations/mottonen.py:164]: I understand that you skip for abstract, but I'm wondering if this is going to be an issue in the future. Using jit we are going to start to have less efficient implementations right?

File: tests/templates/test_subroutines/test_select_pauli_rot.py [hunk 10]
```diff
@@ -130,15 +130,37 @@ def test_correctness(self, angles, rot_axis):
...
         """Test that the correct gates are added in the decomposition"""
 
+        shape = (2**n,) if batch_dim is None else (batch_dim, 2**n)
+        x = np.random.random(shape)
```
Comment (KetpuntoG) [inline tests/templates/test_subroutines/test_select_pauli_rot.py:140]: Could you set a local seed? I think we avoid having non-deterministic tests

**Review by KetpuntoG — COMMENTED** [review]
Great job! It's nice to see such an improvement in performance 😊 I left a couple of comments, once you keep an eye on them, ready to merge

File: pennylane/templates/state_preparations/mottonen.py (hunk 9 above)
Comment (albi3ro) [reply pennylane/templates/state_preparations/mottonen.py:164]: No start.  We've been that for years.

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -120,6 +120,10 @@
 
 <h3>Improvements 🛠</h3>
 
+* Computing the angles for uniformly controlled rotations, used in `MottonenStatePreparation`
+  and `SelectPauliRot`, now takes much less computational effort and memory.
```
Comment (isaacdevlugt) [inline doc/releases/changelog-dev.md:124]:
```suggestion
* Computing the angles for uniformly controlled rotations, used in :class:`~.MottonenStatePreparation`
  and :class:`~.SelectPauliRot`, now takes much less computational effort and memory.
```

File: doc/releases/changelog-dev.md [hunk 12]
```diff
@@ -274,6 +278,9 @@ Here's a list of deprecations made this release. For a more detailed breakdown o
 
 <h3>Bug fixes 🐛</h3>
 
+* Fixes a bug in the validation of `SelectPauliRot` that prevents broadcasting.
```
Comment (isaacdevlugt) [inline doc/releases/changelog-dev.md:281]:
```suggestion
* Fixed a bug in the validation of :class:`~.SelectPauliRot` that prevents parameter broadcasting.
```

File: pennylane/templates/state_preparations/mottonen.py [hunk 13]
```diff
@@ -14,83 +14,107 @@
...
-    return (-1) ** sum_of_ones
+_walsh_hadamard_matrix = np.array([[1, 1], [1, -1]]) / 2
+_cnot_matrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]]).reshape((2,) * 4)
 
 
-def compute_theta(alpha):
-    """Maps the angles alpha of the multi-controlled rotations decomposition of a uniformly controlled rotation
-     to the rotation angles used in the Gray code implementation.
+def compute_theta(alpha: TensorLike, num_qubits: Optional[int] = None):
+    r"""Maps the angles alpha of the multi-controlled rotations decomposition of a
```
Comment (isaacdevlugt) [inline pennylane/templates/state_preparations/mottonen.py:46]:
```suggestion
    r"""Maps the input angles ``alpha`` of the multi-controlled rotations decomposition of a
```

File: pennylane/templates/state_preparations/mottonen.py [hunk 14]
```diff
@@ -14,83 +14,107 @@
...
+    uniformly controlled rotation to the rotation angles used in the Gray code implementation.
```
Comment (isaacdevlugt) [inline pennylane/templates/state_preparations/mottonen.py:47]: Is there a link we can have for "Gray code" ?

File: pennylane/templates/state_preparations/mottonen.py [hunk 15]
```diff
@@ -14,83 +14,107 @@
...
 
 import pennylane as qml
 from pennylane.operation import Operation
+from pennylane.typing import TensorLike
 
 
-# pylint: disable=len-as-condition,arguments-out-of-order,consider-using-enumerate
 def gray_code(rank):
-    """Generates the Gray code of given rank.
+    """Generates the Gray code of given rank, as numeric output.
```
Comment (isaacdevlugt) [inline pennylane/templates/state_preparations/mottonen.py:27]: Should we add a link for Gray code here?

File: pennylane/templates/state_preparations/mottonen.py [hunk 16]
```diff
@@ -14,83 +14,107 @@
...
-        (tensor_like): rotation angles theta
+        tensor_like: The transformed tensor with the same shape as the input ``alpha``.
+
+    Due to the execution of the transform as a sequence of tensor multiplications
+    with shapes ``(2, 2), (2, 2,... 2)->(2, 2,... 2)``, the theoretical scaling of this
+    method is the same as the one for the
+    `Fast Walsh-Hadamard transform <https://en.wikipedia.org/wiki/Fast_Walsh-Hadamard_transform>`__:
+    On :math:`n` qubits, there are :math:`n` calls to ``tensordot``, each multiplying a ``(2, 2)``
+    matrix to a ``(2,)*num_qubits`` vector, with a single axis being contracted. This means
+    that there are :math:`n` operations with a FLOP count of ``4 * 2**(num_qubits-1)``, where ``4``
```
Comment (isaacdevlugt) [inline pennylane/templates/state_preparations/mottonen.py:68]: FLOP = floating point operation?

File: pennylane/templates/state_preparations/mottonen.py (hunk 14 above)
Comment (albi3ro) [reply pennylane/templates/state_preparations/mottonen.py:47]: Or make `compute_theta` and `gray_code` private?

File: pennylane/templates/state_preparations/mottonen.py (hunk 8 above)
Comment (dwierichs) [reply pennylane/templates/state_preparations/mottonen.py:51]: Ah, I just wanted to make clear that we are "applying cnot gates" to something that is just an array of angles, and not a state vector. So "applying a gate" is somewhat wrong but it is what we do computationally :D

File: tests/templates/test_subroutines/test_select_pauli_rot.py (hunk 10 above)
Comment (dwierichs) [reply tests/templates/test_subroutines/test_select_pauli_rot.py:140]: Done via the `seed` fixture :) :heavy_check_mark:

File: pennylane/templates/state_preparations/mottonen.py (hunk 14 above)
Comment (dwierichs) [reply pennylane/templates/state_preparations/mottonen.py]: I think I'd prefer to keep them public because they are not only used in this file (but also in `FABLE`).
Added a link to the wiki page, like in `gray_code` itself. It's fairly well-searchable I'd say :)

File: pennylane/templates/state_preparations/mottonen.py (hunk 16 above)
Comment (dwierichs) [reply pennylane/templates/state_preparations/mottonen.py]: Yes! Introduced the acronym now.

**Review by KetpuntoG — APPROVED** [review]
🚀

**Review by isaacdevlugt — APPROVED** [review]
